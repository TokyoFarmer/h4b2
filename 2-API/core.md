[Home](/) &gt; [@h4bff/core](core.md)

## core package

## Classes

|  Class | Description |
|  --- | --- |
|  [App](core/App.md) | Represents an H4BFF application, the central hub of h4bff. Its the class that loads and initializes all the plugins, storing instances of their singletons in the singleton locator, as well as creating new service contexts which keep per-request service locators. |
|  [AppSingleton](core/AppSingleton.md) |  |
|  [BaseService](core/BaseService.md) |  |
|  [Locator](core/Locator.md) |  |
|  [ServiceContext](core/ServiceContext.md) | Represents a transient context. On the backend that's usually created for every individual HTTP request. On the frontend a transient request is created when the router route changes - when the user navigates to a different page. |
|  [ServiceContextEvents](core/ServiceContextEvents.md) | Handles events related to context creation, destruction etc.<!-- -->See [onContextDisposed](core/ServiceContextEvents.md#oncontextdisposed) for more details. |

## Type Aliases

|  Type Alias | Description |
|  --- | --- |
|  [ConstructorOrFactory](core/ConstructorOrFactory.md) |  |
|  [ContextListener](core/ContextListener.md) |  |
|  [PublicInterface](core/PublicInterface.md) |  |

