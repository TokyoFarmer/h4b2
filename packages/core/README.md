# The Core

Lays out the base for pluggable web application framework.
