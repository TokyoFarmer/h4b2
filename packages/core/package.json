{
  "name": "@h4bff/core",
  "version": "1.0.0",
  "description": "H4BFF, the core: dependency injection, overrides, apps singletons and services.",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "build/**/*"
  ],
  "license": "MIT",
  "scripts": {
    "prepublish": "yarn build",
    "build": "tsc -p tsconfig.json",
    "watch": "tsc -w",
    "test": "jest"
  }
}
