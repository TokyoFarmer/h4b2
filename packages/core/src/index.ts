type ClassFactory<U, T> = (u: U) => T;
type ClassConstructor<U, T> = { new (u: U): T };
export type ConstructorOrFactory<U, T> = ClassFactory<U, T> | ClassConstructor<U, T>;

export type PublicInterface<T> = { [K in keyof T]: T[K] };

export class App {
  singletonLocator: Locator<this>;
  parentApp: App | null;

  /**
   * Represents an H4BFF application.
   *
   * Applications are responsible for:
   * - initializing and storing singletons;
   * - creating and executing functions within service contexes;
   * - managing singleton overrides (useful for tests)
   *
   * Applications can have parents. When initializing singletons,
   * we check the parent, grand-parent app etc. to see if the singleton
   * is already initialized, so we can return that instance. If a parent
   * is not found, the singleton is initialized in the child app, *not*
   * in a parent app.
   */
  constructor(opts: { parentApp?: App } = {}) {
    this.parentApp = opts.parentApp ? opts.parentApp : null;
    this.singletonLocator = new Locator(this, s => '__appSingleton' in s);
  }

  /**
   * @internal
   */
  serviceLocator = new Locator(this.createServiceContext(), s => '__baseService' in s);

  /**
   * Walks through parents and returns an existing singleton instance. Call
   * this method ONLY if you KNOW that the instance exists.
   */
  private getExistingSingleton<T>(Klass: ConstructorOrFactory<App, T>): T {
    if (this.hasOwnSingleton(Klass)) {
      return this.singletonLocator.get(Klass);
    } else {
      if (!this.parentApp) {
        // this should not happen, because prior to calling this method,
        // we check if there's an instance.
        throw new Error('Finished searching through parents but couldnt find the singleton instance');
      }
      return this.parentApp.getExistingSingleton(Klass);
    }
  }

  /**
   * Checks if this App instance has the given singleton initialized
   * in itself.
   */
  private hasOwnSingleton<T>(Klass: ConstructorOrFactory<App, T>) {
    return this.singletonLocator.has(Klass);
  }

  /**
   * Checks if this or any of the parent apps has an instance of the
   * given singleton initialized.
   */
  private hasSingleton<T>(Klass: ConstructorOrFactory<App, T>): boolean {
    if (this.hasOwnSingleton(Klass)) {
      return true;
    } else {
      if (!this.parentApp) {
        return false;
      } else {
        return this.parentApp.hasSingleton(Klass);
      }
    }
  }

  /**
   * Returns an instance of the singleton, if it exists somewhere here or
   * in some of the parent apps. If it doesn't it's created in this app.
   */
  getSingleton<T>(Klass: ConstructorOrFactory<App, T>): T {
    if (this.hasSingleton(Klass)) {
      return this.getExistingSingleton(Klass);
    }
    return this.singletonLocator.get(Klass);
  }

  /**
   * Allows you to specify an alternative implementation for the
   * expected singleton. Each time someone tries to instantiate the
   * specified class / fn, the override is used instead. The type of
   * the override must match that of the original class / fn.
   */
  overrideSingleton<T>(Klass: ConstructorOrFactory<App, T>, Klass2: ConstructorOrFactory<App, PublicInterface<T>>) {
    return this.singletonLocator.override(Klass, Klass2);
  }

  /**
   * Clears any defined singleton overrides.
   */
  clearSingletonOverrides() {
    return this.singletonLocator.clearOverrides();
  }

  /**
   * Allows you to specify an alternative implementation for the
   * expected service. Each time someone tries to instantiate the
   * specified class / fn, the override is used instead. The type of
   * the override must match that of the original class / fn.
   */
  overrideService<T>(
    Klass: ConstructorOrFactory<ServiceContext, T>,
    Klass2: ConstructorOrFactory<ServiceContext, PublicInterface<T>>,
  ) {
    return this.serviceLocator.override(Klass, Klass2);
  }

  /**
   * Clears any defined service overrides.
   */
  clearServiceOverrides() {
    return this.serviceLocator.clearOverrides();
  }

  /**
   * Loads the plugin, which forces its initialization.
   *
   * Use this when you want to initialize a class somewhere
   * specific in the hierarchy of apps, for example in
   * the parent app, to prevent it from being initalized
   * in a child later on.
   */
  load<T>(Klass: ConstructorOrFactory<App, T>): void {
    this.getSingleton(Klass); // force initialization;
    return;
  }

  /**
   * Override this method to load plugins in your app.
   *
   * TODO: describe why it's important to load plugins
   * when configuring an application; also how it differs
   * from starting the application -- the other kind of
   * side-effects.
   */
  loadPlugins() {
    throw new Error('Override this method to load plugins');
  }

  /**
   * Creates a service context.
   *
   * See the documentation for `ServiceContext` for more details.
   */
  createServiceContext() {
    return new ServiceContext(this);
  }

  /**
   * Creates a service context, executes the provided function and disposes
   * of the context afterwards. Disposal happens regardless of exceptions.
   */
  withServiceContext<T>(f: (createdCtx: ServiceContext) => PromiseLike<T>): PromiseLike<T> {
    let serviceContext = this.createServiceContext();
    let ctxEvents = this.getSingleton(ServiceContextEvents);
    let happyHandler = (res: T) => ctxEvents.disposeContext(serviceContext, null).then(() => res);
    let sadHandler = (error: Error) =>
      ctxEvents.disposeContext(serviceContext, error).then(() => {
        throw error;
      });
    try {
      return Promise.resolve(f(serviceContext)).then(happyHandler, sadHandler);
    } catch (e) {
      return sadHandler(e);
    }
  }

  /**
   * When instatiating singletons, child applications look in their parents
   * for already instantiated singletons, returning them if they exists.
   *
   * Services and the service context are not affected by parent / child
   * hierarchies.
   *
   * Use this when you want to initialize the same kind of a singleton
   * multiple times.
   */
  createChildApp() {
    return new App({ parentApp: this });
  }
}

export class AppSingleton {
  protected static __appSingleton = true;

  /**
   * Derive from this class to create application singletons.
   *
   * Singletons are initialized only once per application, although
   * you can initialize different singletons of the same type in
   * child applications.
   */
  constructor(protected app: App) {}

  /**
   * A proxy for `app.getSingleton(Klass)`.
   */
  getSingleton<T>(Klass: ConstructorOrFactory<App, T>): T {
    return this.app.getSingleton(Klass);
  }
}

export class ServiceContext {
  private _locator: Locator<ServiceContext> | null = null;

  get locator() {
    if (this._locator == null) {
      this._locator = this.app.serviceLocator.withNewContext(this);
    }
    return this._locator;
  }

  /**
   * Represents a transient context, that's usually created when an
   * HTTP request comes, a new page is created or on similar events
   * on which you want to create and later destroy some services.
   *
   * For example, if you want all services to exectute their queries
   * within a single transaction, you'd create a service context and
   * initialize a transaction inside it.
   *
   * Another example is handling access to the request that triggered
   * the creation of the service context. You'll want to initialize
   * a Request service on the service context which will get passed
   * around.
   *
   * See `ServiceContextEvents` for more info on doing things in
   * response to the creation / destruction of service contexes.
   */
  constructor(private app: App) {}

  /**
   * Initializes the class within the service context (itself.)
   *
   * If the service is already initialized, it returns the instance.
   */
  getService<T extends BaseService>(SvcClass: ConstructorOrFactory<ServiceContext, T>): T {
    return this.locator.get(SvcClass);
  }

  /**
   * A proxy for `app.getSingleton(Klass)`.
   */
  getSingleton<T>(SingletonClass: ConstructorOrFactory<App, T>): T {
    return this.app.getSingleton(SingletonClass);
  }
}

export type ContextListener = (serviceCtx: ServiceContext, error: Error | null) => PromiseLike<void>;

/**
 * Handles events related to context creation, destruction etc.
 *
 * See `onContextDisposed` for more details.
 */
export class ServiceContextEvents extends AppSingleton {
  private listeners: ContextListener[] = [];

  /**
   * Registers a function that will be called every time a service
   * context is getting destroyed.
   *
   * For example, let's say we create a context on the backend
   * each time a request comes in. Before we return a response,
   * we want to close the DB transaction. What you'd do is:
   *
   * `app.getSingleton(ServiceContextEvents).onContextDisposed(ctx => { ... ctx.getService(Txn).rm() })`
   *
   * Use this when you want to react to the destruction of any
   * service context.
   */
  onContextDisposed(listener: ContextListener) {
    this.listeners.push(listener);
  }

  /**
   * Triggers the destruction of the passed service context.
   */
  disposeContext: ContextListener = (serviceCtx, err) => {
    return Promise.all(this.listeners.map(l => l(serviceCtx, err))).then(() => {});
  };
}

export class BaseService {
  protected static __baseService = true;
  static _factory: any;
  static get factory() {
    if (!this._factory) this._factory = (serviceCtx: ServiceContext) => new this(serviceCtx);
    return this._factory;
  }

  /**
   * Derive from this class to create H4BFF services.
   *
   * Services are classes that are instantiated and operate
   * within an "isolated" service context, and are instantiated
   * separately within each context, as opposed to singletons
   * which have only one instance within an App.
   *
   * Examples of classes that should derive from `BaseService`:
   * * Request: holds a reference to the HTTP request that triggered
   *   the service context creation
   * * Transaction: a single transaction that's shared between
   *   services that operate througout the duration of a single
   *   request.
   * * UserInfo: information about the current user (known through
   *   req.session.id)
   * etc.
   */
  constructor(protected context: ServiceContext) {}

  /**
   * Proxy for `serviceContext.getService(Klass)`.
   */
  getService<T extends BaseService>(SvcClass: { new (sc: ServiceContext): T }): T {
    return this.context.getService(SvcClass);
  }

  /**
   * Proxy for `app.getSingleton(Klass)`.
   */
  getSingleton<T>(SingletonClass: ConstructorOrFactory<App, T>): T {
    return this.context.getSingleton(SingletonClass);
  }
}

export class Locator<Context> {
  private instances: Map<Function, any> = new Map();
  private overrides: Map<Function, Function> = new Map();

  /**
   * Mostly for internal use.
   *
   * Used for locating instances of classes / factory functions,
   * and for instantiating them if they don't exist.
   *
   * Provides overriding functionality (used for mocks in tests.)
   */
  constructor(
    private locatorCtx: Context,
    private isClass: (v: ConstructorOrFactory<Context, any>) => boolean,
    options?: {
      overrides?: Map<Function, Function>;
    },
  ) {
    options = options || {};
    this.overrides = options.overrides || this.overrides;
  }

  private isClassTG<T>(v: ConstructorOrFactory<Context, T>): v is ClassConstructor<Context, T> {
    return this.isClass(v);
  }

  private instantiate<T>(f: ConstructorOrFactory<Context, T>) {
    if (this.isClassTG(f)) return new f(this.locatorCtx);
    else return f(this.locatorCtx);
  }

  has<T>(f: ConstructorOrFactory<Context, T>): boolean {
    return this.instances.has(f) || this.overrides.has(f);
  }

  get<T>(f: ConstructorOrFactory<Context, T>): T {
    if (this.overrides.has(f)) {
      f = this.overrides.get(f) as any;
    }
    if (!this.instances.has(f)) {
      this.instances.set(f, this.instantiate(f));
    }
    return this.instances.get(f) as T;
  }

  override<T>(f: ConstructorOrFactory<Context, T>, g: ConstructorOrFactory<Context, T>) {
    if (this.instances.has(f)) {
      console.warn(`Warning: by overriding ${f.name}, you will be shadowing an already instantiated class.`);
    }
    this.overrides.set(f, g);
  }

  withNewContext(ctx: Context) {
    return new Locator(ctx, this.isClass, { overrides: this.overrides });
  }

  clearOverrides() {
    this.overrides = new Map();
  }
}
