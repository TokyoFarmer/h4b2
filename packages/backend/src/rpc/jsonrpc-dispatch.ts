import { AppSingleton, ServiceContext } from '@h4bff/core';
import { ErrorResponse } from './error-handler';
import { isCustomResponse } from './response';
import { Response, Request } from 'express';
import { RequestContextProvider, RequestInfo } from '../request';
import { RPCDispatcher, DispatchInfo } from './dispatcher';

/**
 * This class allows you to expose the RPC service registry as a JSON-RPC api endpoint.
 *
 * To install, simply use JSONRPCDispatch's routeHandler and mount it at a specific express
 * route e.g. `/api/v1`.
 */
export class JSONRPCDispatch extends AppSingleton {
  /**
   * @internal
   */
  getJSONRPCDispatchInfo(ctx: ServiceContext): () => DispatchInfo {
    return () => {
      let req = ctx.getService(RequestInfo).req;

      const rpcPath = req.query && req.query.method;
      const params = req.body && req.body.params;
      if (!rpcPath) return { params, method: null, service: null };

      const lastDotIndex = rpcPath && rpcPath.lastIndexOf('.');
      const [service, method] = [rpcPath.slice(0, lastDotIndex), rpcPath.slice(lastDotIndex + 1)];

      return { service, method, params };
    };
  }

  private jsonFail(res: Response, code: number, message: string, data: any = null) {
    return res.status(code).json({
      code,
      result: data,
      error: {
        code,
        message,
      },
      version: 2,
      backendError: true,
    });
  }

  private fail = (res: Response, e: Error | ErrorResponse) => {
    if (e != null && (e as any).data != null && 'data' in e) {
      return this.jsonFail(res, e.code, e.message, e.data);
    }

    if (typeof (e as any).code === 'number') {
      return this.jsonFail(res, (e as any).code, e.message);
    }

    console.error(e);
    return this.jsonFail(res, 500, 'An unexpected error occurred. Please try again.');
  };

  private success = (res: Response, data: any, code: number = 200) => {
    if (isCustomResponse(data)) {
      return data.sendToHTTPResponse(res, code);
    } else {
      return res.status(code).json({
        code,
        result: data,
        error: null,
        version: 2,
      });
    }
  };

  /**
   * Middleware that adds RPC handling for given request and response.
   * It binds the request and response to a service context and forwards the
   * request to the {@link RPCDispatcher}.
   */
  routeHandler = (req: Request, res: Response) => {
    return this.getSingleton(RequestContextProvider)
      .withRequestContext(req, res, ctx =>
        ctx
          .getService(RPCDispatcher)
          .withDispatchInfo(this.getJSONRPCDispatchInfo(ctx))
          .call(),
      )
      .then(data => this.success(res, data, 200), err => this.fail(res, err));
  };
}
