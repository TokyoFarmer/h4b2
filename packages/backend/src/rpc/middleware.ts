import * as Promise from 'bluebird';
import { AppSingleton } from '@h4bff/core';
import { RPCDispatcher } from './dispatcher';

export type RPCMiddleware = (dispatcher: RPCDispatcher, next: () => Promise<any>) => Promise<any>;

/**
 * Container for RPC middlewares.
 */
export class RPCMiddlewareContainer extends AppSingleton {
  call = (dispatcher: RPCDispatcher) => {
    return Promise.resolve().then(() => dispatcher.handleRequest());
  };

  /**
   * Adds new middleware to the RPC layer. Make sure that a middleware always returns a result.
   * It might be a result that the middleware generates or modifies from the previos middleware,
   * or it could be the result of the "next()" middleware. This way the result will be cascaded
   * to the end of the middleware chain to achieve consistent success and error handling.
   */
  addMiddleware(middleware: RPCMiddleware) {
    let oldCall = this.call;
    this.call = (dispatcher: RPCDispatcher) => {
      try {
        return middleware.call(null, dispatcher, () => oldCall(dispatcher));
      } catch (e) {
        return Promise.reject(e);
      }
    };
  }
}
