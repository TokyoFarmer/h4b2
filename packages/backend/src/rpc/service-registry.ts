import { BaseService, AppSingleton, ServiceContext } from '@h4bff/core';
import { JSONRPCExpress } from './jsonrpc-express';

/**
 * RPC service middleware.
 */
export type RPCServiceMiddleware = (
  sCtx: ServiceContext,
  next: () => PromiseLike<void>,
) => PromiseLike<void>;

/**
 * Responsible for holding the RPC service mapping.
 */
export class RPCServiceRegistry extends AppSingleton {
  services: { [key: string]: typeof BaseService } = {};

  /**
   * Adds new RPC service mapping.
   * @param alias - service alias
   * @param service - service constructor
   */
  add(alias: string, service: typeof BaseService) {
    if (this.services[alias] != null) {
      throw new Error('Namespace ' + alias + ' already in use!');
    }
    this.services[alias] = service;
  }

  /**
   * Returns service for given alias.
   */
  get(serviceAlias: string): typeof BaseService | undefined {
    return this.services[serviceAlias];
  }

  /**
   * @deprecated: Use JSONRRPCExpress.routeHandler instead
   */
  routeHandler = this.getSingleton(JSONRPCExpress).routeHandler;
}
