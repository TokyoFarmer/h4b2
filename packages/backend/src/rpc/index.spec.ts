import { App, BaseService, ServiceContextEvents } from '@h4bff/core';
import { Request, Response } from 'express';
import { RPCServiceRegistry } from './service-registry';
import { RPCMiddlewareContainer } from './middleware';
import { RPCErrorHandlers } from '.';
import { JSONRPCDispatch } from './jsonrpc-dispatch';

function mockRequest(method?: string, params?: any) {
  const req = {} as Request;
  req.query = { method };
  req.body = { params };
  return req;
}

function mockResponse() {
  const res = {} as Response;
  res.status = jest.fn(() => res);
  res.json = jest.fn();
  return res;
}

function runMock(
  app: App,
  opts: {
    method?: string;
    params?: any;
    response?: (params: unknown) => any;
    req?: Request;
    res?: Response;
  },
) {
  let sendResponse = opts.response || (() => 'Hello World');
  let mReq = opts.req || mockRequest(opts.method, opts.params);
  let mRes = opts.res || mockResponse();

  let disposed = false;
  app.getSingleton(ServiceContextEvents).onContextDisposed(async () => {
    await new Promise(resolve => setTimeout(resolve, 1));
    disposed = true;
  });

  app.getSingleton(RPCServiceRegistry).add(
    'test',
    class TestService extends BaseService {
      method(params: unknown) {
        return sendResponse(params);
      }
    },
  );

  return app
    .getSingleton(JSONRPCDispatch)
    .routeHandler(mReq, mRes)

    .then(() => {
      expect(disposed).toBe(true);
      return { req: mReq, res: mRes };
    });
}

describe('RPCDispatcher', () => {
  describe('handleRequest', () => {
    it(`should respond with error if method is not found in the request query params`, async () => {
      let { res } = await runMock(new App(), {});
      expect(res.status).toHaveBeenCalledWith(400);
      expect(res.json).toHaveBeenCalledWith({
        code: 400,
        result: null,
        error: {
          code: 400,
          message: '"method" query parameter not found',
        },
        version: 2,
        backendError: true,
      });
    });

    it(`should respond with error if body params are not found in the request`, async () => {
      let { res } = await runMock(new App(), { method: 'test.method' });

      expect(res.status).toHaveBeenCalledWith(400);
      expect(res.json).toHaveBeenCalledWith({
        code: 400,
        result: null,
        error: {
          code: 400,
          message: '"params" not found, send an empty object in case of no parameters',
        },
        version: 2,
        backendError: true,
      });
    });

    it(`should respond with error if method doesn't exist on the service`, async () => {
      let requestInfo = await runMock(new App(), { method: 'test.notMethod', params: {} });

      expect(requestInfo.res.status).toHaveBeenCalledWith(404);
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 404,
        result: null,
        error: {
          code: 404,
          message: 'Method not found',
        },
        version: 2,
        backendError: true,
      });
    });
  });

  describe('RPC call execution', () => {
    it('should respond with success and dispose context in normal circumstances', async () => {
      let app = new App();

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => 'data',
      });
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 200,
        result: 'data',
        error: null,
        version: 2,
      });
      expect(requestInfo.res.status).toHaveBeenCalledWith(200);
    });

    it('should respond with custom success response and dispose context for specific results', async () => {
      let app = new App();

      // mock middleware call and prepare custom response
      let mockData = {
        sendToHTTPResponse: jest.fn(() => Promise.resolve('test')),
      };

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => mockData,
      });
      expect(mockData.sendToHTTPResponse).toHaveBeenCalledWith(requestInfo.res, 200);
    });

    it('should respond with error response when the RPC call fails', async () => {
      let app = new App();

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => Promise.reject({ code: 403, message: 'error' }),
      });

      expect(requestInfo.res.status).toHaveBeenCalledWith(403);
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 403,
        result: null,
        error: {
          code: 403,
          message: 'error',
        },
        version: 2,
        backendError: true,
      });
    });

    it('should respond with error response from a registered error handler when the RPC call fails with error the handler can handle', async () => {
      let app = new App();

      // create dummy handler to ensure the correct handler can be reached
      let dummyHandler = jest.fn((_e: Error) => undefined);

      // create custom handler that we expect to be reached
      let specificHandler = jest.fn((e: Error) => {
        if ((e as any).isJoi) {
          return {
            code: 400,
            message: 'Technical error, the request was malformed.',
            data: (e as any).details,
          };
        }
      });

      // register the handlers in order
      app.getSingleton(RPCErrorHandlers).addErrorHandler(dummyHandler);
      app.getSingleton(RPCErrorHandlers).addErrorHandler(specificHandler);

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => Promise.reject({ isJoi: true, details: 'Error details' }),
      });
      // expect(err).toEqual(
      // expect.objectContaining({ isJoi: true, details: 'Error details' }),
      // );
      expect(requestInfo.res.status).toHaveBeenCalledWith(400);
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 400,
        result: 'Error details',
        error: {
          code: 400,
          message: 'Technical error, the request was malformed.',
        },
        version: 2,
        backendError: true,
      });

      expect(dummyHandler).not.toHaveBeenCalled();
      expect(specificHandler).toHaveBeenCalledTimes(1);
    });

    it('should respond with error response from the last registered error handler when the RPC call fails with error that the handler can handle', async () => {
      let app = new App();

      // create handler that can handle the error but cannot be reached? reached
      let specificHandler1 = jest.fn((e: Error) => {
        if ((e as any).isJoi) {
          return {
            code: 400,
            message: 'Technical error, the request was malformed.',
            data: (e as any).details,
          };
        }
      });

      // create third handler with same condition as the previous to ensure it was not reached
      let specificHandler2 = jest.fn((e: Error) => {
        if ((e as any).isJoi) {
          return {
            code: 444,
            message: 'I am the most recently registered handler and I am handling the error.',
            data: (e as any).details,
          };
        }
      });

      // register the handlers in order
      app.getSingleton(RPCErrorHandlers).addErrorHandler(specificHandler1);
      app.getSingleton(RPCErrorHandlers).addErrorHandler(specificHandler2);

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => Promise.reject({ isJoi: true, details: 'Error details' }),
      });
      // expect(err).toEqual(
      //   expect.objectContaining({ isJoi: true, details: 'Error details' }),
      // );
      expect(requestInfo.res.status).toHaveBeenCalledWith(444);
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 444,
        result: 'Error details',
        error: {
          code: 444,
          message: 'I am the most recently registered handler and I am handling the error.',
        },
        version: 2,
        backendError: true,
      });
      expect(specificHandler1).not.toHaveBeenCalledWith();
      expect(specificHandler2).toHaveBeenCalledTimes(1);
    });

    it('should respond with error response from the default coded error handler if the RPC call fails with error no handler can handle', async () => {
      let app = new App();

      // create custom handler that we expect to be reached
      let specificHandler = jest.fn((e: Error) => {
        if ((e as any).isJoi) {
          return {
            code: 400,
            message: 'Technical error, the request was malformed.',
            data: (e as any).details,
          };
        }
      });

      // register the handler
      app.getSingleton(RPCErrorHandlers).addErrorHandler(specificHandler);

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => Promise.reject({ code: 444, message: 'Error details' }),
      });

      // expect(err).toEqual(expect.objectContaining({ code: 444, message: 'Error details' }));
      expect(requestInfo.res.status).toHaveBeenCalledWith(444);
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 444,
        result: null,
        error: {
          code: 444,
          message: 'Error details',
        },
        version: 2,
        backendError: true,
      });

      expect(specificHandler).toHaveBeenCalledTimes(1);
    });

    it('should respond with 500 error response when the RPC call fails with unknown error', async () => {
      let app = new App();
      // prepare request / response

      let requestInfo = await runMock(app, {
        method: 'test.method',
        params: {},
        response: () => Promise.reject('error'),
      });

      // expect(err).toEqual('error');
      expect(requestInfo.res.status).toHaveBeenCalledWith(500);
      expect(requestInfo.res.json).toHaveBeenCalledWith({
        code: 500,
        result: null,
        error: {
          code: 500,
          message: 'An unexpected error occurred. Please try again.',
        },
        version: 2,
        backendError: true,
      });
    });
  });
});

async function runLifecycleTest(opts?: {
  mw1fail?: 'before' | 'after';
  mw2fail?: 'before' | 'after';
  methodfail?: boolean;
  disposefail?: boolean;
}) {
  let lifecycle = ['start'];
  let app = new App();
  app.getSingleton(RPCServiceRegistry).add(
    'x',
    class X extends BaseService {
      method(params: { test: string }) {
        lifecycle.push(params.test);
        if (opts && opts.methodfail) throw new Error('method fail');
        return Promise.resolve('Hello world');
      }
    },
  );
  app.getSingleton(RPCMiddlewareContainer).addMiddleware(async (_disp, next) => {
    lifecycle.push('middleware1 before');
    if (opts && opts.mw1fail === 'before') throw new Error('mw1 fail before');
    try {
      let res = await next();
      lifecycle.push('middleware1 after');
      if (opts && opts.mw1fail === 'after') throw new Error('mw1 fail after');
      return res;
    } catch (e) {
      lifecycle.push('middleware1 after-error');
      throw e;
    }
  });

  app.getSingleton(RPCMiddlewareContainer).addMiddleware(async (_disp, next) => {
    lifecycle.push('middleware2 before');
    if (opts && opts.mw2fail === 'before') throw new Error('mw2 fail before');
    let res = await next();
    lifecycle.push('middleware2 after');
    if (opts && opts.mw2fail === 'after') throw new Error('mw2 fail after');
    return res;
  });
  app.getSingleton(ServiceContextEvents).onContextDisposed(async () => {
    lifecycle.push('context disposing');
    if (opts && opts.disposefail) throw new Error('dispose fail');
    await new Promise(resolve => setTimeout(resolve, 1));
    lifecycle.push('context fully disposed');
  });

  app.getSingleton(RPCErrorHandlers).addErrorHandler(_e => {
    lifecycle.push('error handler called');
    return undefined;
  });

  let res = mockResponse();
  res.status = code => {
    lifecycle.push('sendCode:' + code.toString());
    return res;
  };
  res.json = json => {
    lifecycle.push('sendJson:' + JSON.stringify(json));
    return res;
  };

  await runMock(app, { method: 'x.method', params: { test: 'method' }, res });

  return lifecycle;
}
describe('rpc lifecycle', () => {
  let variants = [
    { mw1fail: 'before' as const },
    { mw1fail: 'after' as const },
    { mw2fail: 'before' as const },
    { mw2fail: 'after' as const },
    { disposefail: true },
    { methodfail: true },
    undefined,
  ];
  for (let combo of variants) {
    let comboStr = JSON.stringify(combo);
    it('should work in the combination ' + comboStr, async () => {
      let lifecycle = await runLifecycleTest(combo);
      expect(lifecycle).toMatchSnapshot();
    });
  }
});
