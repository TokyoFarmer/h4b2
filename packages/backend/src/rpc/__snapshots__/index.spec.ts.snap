// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rpc lifecycle should work in the combination {"disposefail":true} 1`] = `
Array [
  "start",
  "middleware2 before",
  "middleware1 before",
  "method",
  "middleware1 after",
  "middleware2 after",
  "context disposing",
  "sendCode:500",
  "sendJson:{\\"code\\":500,\\"result\\":null,\\"error\\":{\\"code\\":500,\\"message\\":\\"An unexpected error occurred. Please try again.\\"},\\"version\\":2,\\"backendError\\":true}",
]
`;

exports[`rpc lifecycle should work in the combination {"methodfail":true} 1`] = `
Array [
  "start",
  "middleware2 before",
  "middleware1 before",
  "method",
  "middleware1 after-error",
  "error handler called",
  "context disposing",
  "context fully disposed",
  "sendCode:500",
  "sendJson:{\\"code\\":500,\\"result\\":null,\\"error\\":{\\"code\\":500,\\"message\\":\\"An unexpected error occurred. Please try again.\\"},\\"version\\":2,\\"backendError\\":true}",
]
`;

exports[`rpc lifecycle should work in the combination {"mw1fail":"after"} 1`] = `
Array [
  "start",
  "middleware2 before",
  "middleware1 before",
  "method",
  "middleware1 after",
  "middleware1 after-error",
  "error handler called",
  "context disposing",
  "context fully disposed",
  "sendCode:500",
  "sendJson:{\\"code\\":500,\\"result\\":null,\\"error\\":{\\"code\\":500,\\"message\\":\\"An unexpected error occurred. Please try again.\\"},\\"version\\":2,\\"backendError\\":true}",
]
`;

exports[`rpc lifecycle should work in the combination {"mw1fail":"before"} 1`] = `
Array [
  "start",
  "middleware2 before",
  "middleware1 before",
  "error handler called",
  "context disposing",
  "context fully disposed",
  "sendCode:500",
  "sendJson:{\\"code\\":500,\\"result\\":null,\\"error\\":{\\"code\\":500,\\"message\\":\\"An unexpected error occurred. Please try again.\\"},\\"version\\":2,\\"backendError\\":true}",
]
`;

exports[`rpc lifecycle should work in the combination {"mw2fail":"after"} 1`] = `
Array [
  "start",
  "middleware2 before",
  "middleware1 before",
  "method",
  "middleware1 after",
  "middleware2 after",
  "error handler called",
  "context disposing",
  "context fully disposed",
  "sendCode:500",
  "sendJson:{\\"code\\":500,\\"result\\":null,\\"error\\":{\\"code\\":500,\\"message\\":\\"An unexpected error occurred. Please try again.\\"},\\"version\\":2,\\"backendError\\":true}",
]
`;

exports[`rpc lifecycle should work in the combination {"mw2fail":"before"} 1`] = `
Array [
  "start",
  "middleware2 before",
  "error handler called",
  "context disposing",
  "context fully disposed",
  "sendCode:500",
  "sendJson:{\\"code\\":500,\\"result\\":null,\\"error\\":{\\"code\\":500,\\"message\\":\\"An unexpected error occurred. Please try again.\\"},\\"version\\":2,\\"backendError\\":true}",
]
`;

exports[`rpc lifecycle should work in the combination undefined 1`] = `
Array [
  "start",
  "middleware2 before",
  "middleware1 before",
  "method",
  "middleware1 after",
  "middleware2 after",
  "context disposing",
  "context fully disposed",
  "sendCode:200",
  "sendJson:{\\"code\\":200,\\"result\\":\\"Hello world\\",\\"error\\":null,\\"version\\":2}",
]
`;
